streamingccl: add new logical package

stevendanna opened the pull request:
This adds a new LogicalReplication job and LogicalReplicationWriter
processor. This is a stub implementation that will be changed
substantially in the coming weeks.

The idea of the LogicalReplication job is that it can be used to
ingest data from one cluster into another while both clusters are
still accepting writes.

When two LogicalReplication jobs are started operating in opposite
directions on a pair of servers, the end effect will be an
active-active asynchronous replication process.

The overall architecture is similar to physical cluster replication.

Namely, for one flow, the destination cluster asks the source cluster
to establish a protected timestamp over the requested rows and to
return an ideal ingestion plan to the source.

The source then starts a set of workers each of which is responsible
for streaming a subset of the source cluster's keyspace. Rows are
returned over a pgwire connection.

The implementation here is contains almost none of the pieces required
to make this production ready.

It is missing, among many other things:

  1. Robust handling of schema descriptors to ensure that the user
  cannot make incompatible schema changes during replication.

  2. Correct handling of table's related by FK constraints.

  3. Easy-to-use SQL syntax.

  4. Correct last-write-wins semantics for DELETEs.

  5. Last-write-wins semantics without changes to the user's schema.

The goal of this PR is simply to get the job and processor in place so
that we can start filling out the production implementation.

Epic: none
Release note: None

stevendanna committed:
streamingccl: add new logical package

This adds a new LogicalReplication job and LogicalReplicationWriter
processor. This is a stub implementation that will be changed
substantially in the coming weeks.

The idea of the LogicalReplication job is that it can be used to
ingest data from one cluster into another while both clusters are
still accepting writes.

When two LogicalReplication jobs are started operating in opposite
directions on a pair of servers, the end effect will be an
active-active asynchronous replication process.

The overall architecture is similar to physical cluster replication.

Namely, for one flow, the destination cluster asks the source cluster
to establish a protected timestamp over the requested rows and to
return an ideal ingestion plan to the source.

The source then starts a set of workers each of which is responsible
for streaming a subset of the source cluster's keyspace. Rows are
returned over a pgwire connection.

The implementation here is contains almost none of the pieces required
to make this production ready.

It is missing, among many other things:

  1. Robust handling of schema descriptors to ensure that the user
  cannot make incompatible schema changes during replication.

  2. Correct handling of table's related by FK constraints.

  3. Easy-to-use SQL syntax.

  4. Correct last-write-wins semantics for DELETEs.

  5. Last-write-wins semantics without changes to the user's schema.

The goal of this PR is simply to get the job and processor in place so
that we can start filling out the production implementation.

Epic: none
Release note: None

Co-authored-by: Jeremy Yang <jyang@cockroachlabs.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124606)
<!-- Reviewable:end -->

msbutler reviewed (approved):
Looks production ready to me. Go ahead and merge now and address comments later if you'd like!

stevendanna commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -207,6 +207,52 @@ message HistoryRetentionProgress {
+    // through the lifetime of a replication stream. This will be the timestamp as
+    // of which each partition will perform its initial rangefeed scan on the
+    // source cluster.
+    util.hlc.Timestamp replication_start_time = 3 [(gogoproto.nullable) = false];
```
The second is the rule of thumb I was going by. When I move initial contact to planning, I'll move where these live.

stevendanna committed:
kvserver: bump max range size in TestStoreRangeSplitBackpressureWrites

Similar to #124818, this test assumes that no other ranges will
ever split but the max range size was very close to the size of an
existing system range.

Epic: none
Release note: None

stevendanna commented:
bors r=msbutler

nicktrav commented:
blathers backport 24.1

nicktrav commented:
Ignore my derping 😂 ... wrong window.
